examples/protonet/main.py

```diff
@@ -0,0 +1,105 @@
+"""
+This example is about training prototypical networks to perform N-Way-K-Shot problems.
+
+Reference:
+    Snell, J., Swersky, K. and Zemel, R., 2017.
+    Prototypical networks for few-shot learning.
+    Advances in neural information processing systems, 30.
+"""
+import argparse
+import os
+
+# import sys
+# sys.path.append("/home/wenrui/Projects/pykale/")
+from datetime import datetime
+from typing import Any
+
+import pytorch_lightning as pl
+import torch
+from config import get_cfg_defaults
+from torch.utils.data import DataLoader
+from torchvision import transforms
+from torchvision.models import *
+
+from kale.embed.image_cnn import *
+from kale.loaddata.n_way_k_shot import NWayKShotDataset
+from kale.pipeline.protonet import ProtoNetTrainer
+
+
+def get_parser():
+    parser = argparse.ArgumentParser(description="Args of ProtoNet")
+    parser.add_argument("--cfg", default="examples/protonet/configs/omniglot_resnet18_5way5shot.yaml", type=str)
+    parser.add_argument("--gpus", default=1, type=int)
+    parser.add_argument("--ckpt", default=None, type=Any)
+    return parser
+
+
+def main():
+    # ---- get args ----
+    parser = get_parser()
+    args = parser.parse_args()
+    args.gpus = min(args.gpus, torch.cuda.device_count())
+
+    # ---- get configurations ----
+    cfg_path = args.cfg
+    cfg = get_cfg_defaults()
+    cfg.merge_from_file(cfg_path)
+    cfg.DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
+    cfg.freeze()
+
+    # ---- set model ----
+    net = eval(f"{cfg.MODEL.BACKBONE}(weights={cfg.MODEL.PRETRAIN_WEIGHTS})")
+    if cfg.MODEL.BACKBONE.startswith("resnet"):
+        net.fc = Flatten()
+    model = ProtoNetTrainer(cfg=cfg, net=net)
+
+    # ---- set data loader ----
+    transform = transforms.Compose(
+        [transforms.Resize((cfg.DATASET.IMG_SIZE, cfg.DATASET.IMG_SIZE)), transforms.ToTensor()]
+    )
+    train_set = NWayKShotDataset(
+        path=cfg.DATASET.ROOT,
+        mode="train",
+        k_shot=cfg.TRAIN.K_SHOTS,
+        query_samples=cfg.TRAIN.K_QUERIES,
+        transform=transform,
+    )
+    train_dataloader = DataLoader(train_set, batch_size=cfg.TRAIN.N_WAYS, shuffle=True, drop_last=True)
+    val_set = NWayKShotDataset(
+        path=cfg.DATASET.ROOT, mode="val", k_shot=cfg.VAL.K_SHOTS, query_samples=cfg.VAL.K_QUERIES, transform=transform
+    )
+    val_dataloader = DataLoader(val_set, batch_size=cfg.VAL.N_WAYS, drop_last=True)
```

Comment:
The drop_last=True is an essential set to follow the N-way-K-shot protocol. Because the size of input features in loss function module is strictly pre-defined by N-way-K-shot protocol, if there are only 4 classes for a 5w5s problem, it will raise error.